include/pybind11/cast.h

```diff
@@ -1637,10 +1637,15 @@ str_attr_accessor object_api<D>::attr_with_type_hint(const char *key) const {
         throw std::runtime_error("__annotations__[\"" + std::string(key) + "\"] was set already.");
     }
 
-    const char *text = make_caster<T>::name.text;
+    static constexpr auto type_name = make_caster<T>::name;
+    PYBIND11_DESCR_CONSTEXPR auto types = decltype(type_name)::types();
 
+    const char *text = type_name.text;
+
+    size_t type_index = 0;
     size_t unused = 0;
```

Comment:
Could you please make this
```
    size_t unused_type_index = 0;
    size_t unsued_arg_index = 0;
```
Please also remove the newly added empty line after these, because the variables are very tightly coupled to the function call.